src/nccl/safe.rs

```diff
@@ -0,0 +1,271 @@
+        recvbuff: &mut CudaSlice<T>,
+    ) -> Result<result::NcclStatus, result::NcclError> {
+        unsafe {
+            result::all_gather(
+                sendbuff.cu_device_ptr as *mut _,
+                recvbuff.cu_device_ptr as *mut _,
+                sendbuff.len,
+                T::as_nccl_type(),
+                self.comm,
+                self.device.stream as *mut _,
+            )
+        }
+    }
+
+    pub fn all_reduce<T: NcclType>(
+        &self,
+        sendbuff: &CudaSlice<T>,
+        recvbuff: &mut CudaSlice<T>,
+        reduce_op: &ReduceOp,
+    ) -> Result<result::NcclStatus, result::NcclError> {
+        unsafe {
+            result::all_reduce(
+                sendbuff.cu_device_ptr as *mut _,
+                recvbuff.cu_device_ptr as *mut _,
+                sendbuff.len,
+                T::as_nccl_type(),
+                convert_to_nccl_reduce_op(&reduce_op),
+                self.comm,
+                self.device.stream as *mut _,
+            )
+        }
+    }
+
+    pub fn reduce<T: NcclType>(
+        &self,
+        sendbuff: &CudaSlice<T>,
+        recvbuff: &mut CudaSlice<T>,
+        reduce_op: &ReduceOp,
+        root: i32,
+    ) -> Result<result::NcclStatus, result::NcclError> {
+        unsafe {
+            result::reduce(
+                sendbuff.cu_device_ptr as *mut _,
+                recvbuff.cu_device_ptr as *mut _,
+                sendbuff.len,
+                T::as_nccl_type(),
+                convert_to_nccl_reduce_op(&reduce_op),
+                root,
+                self.comm,
+                self.device.stream as *mut _,
+            )
+        }
+    }
+
+    pub fn reduce_scatter<T: NcclType>(
+        &self,
+        sendbuff: &CudaSlice<T>,
+        recvbuff: &mut CudaSlice<T>,
+        reduce_op: &ReduceOp,
+    ) -> Result<result::NcclStatus, result::NcclError> {
+        unsafe {
+            result::reduce_scatter(
+                sendbuff.cu_device_ptr as *mut _,
+                recvbuff.cu_device_ptr as *mut _,
+                recvbuff.len,
+                T::as_nccl_type(),
+                convert_to_nccl_reduce_op(&reduce_op),
+                self.comm,
+                self.device.stream as *mut _,
+            )
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! group {
+    ($x:block) => {
+        unsafe {
+            result::group_start().unwrap();
+        }
+        $x
+        unsafe {
+            result::group_end().unwrap();
+        }
+    };
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_all_reduce() {
+        let mut threads = vec![];
+        let n = 2;
+        let n_devices = CudaDevice::count().unwrap() as usize;
+        let id = Id::new().unwrap();
+        for i in 0..n_devices {
+            threads.push(std::thread::spawn(move || {
+                let dev = CudaDevice::new(i).unwrap();
+                let comm = Comm::from_rank(dev.clone(), n_devices, id).unwrap();
```

Comment:
I'm wondering if we should force creationg Comm's via some sort of `Vec<CudaDevice>` to avoid errors with id/n_devices. Something like:

```rust
fn init_group(devices: &[Arc<CudaDevice>]) -> Vec<Comm> { ... }

let devs: Vec<CudaDevice> = ...;
let comms: Vec<Comm> = init_group(&devs);
for comm in comms.drain(..) {
    std::thread::spawn(move || { ... }
}
```

Reply:
The group must be initialized from different threads (the call is blocking).
We could do your solution with `group_start/group_end`, but I'm not sure how if sending the comm is actually thread safe (in addition with the cuda context issue).

This is not exactly clear from the docs: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#ncclcomminitrank

vs 

https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#c.ncclCommInitAll

and multiple groups might be initialized. (for the id).

From rank might also be called by different processes. 

If you're Ok I'll try to enable rusty versions of all those various calls (masking groupStart).